### ISSUE TITLE
TLS endpoint wrappers

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5642]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5642|
|Type|enhancement|
|Created|2012-04-12 19:49:43Z|
|Branch|https://github.com/twisted/twisted/tree/tls-endpoint-wrapper-5642-5|

If I construct my own stream client or stream server endpoint, in addition to whatever interesting routing or shuttling-around of I'm doing with the bytes, I might want to encrypt them because they go over some public network as part of the process.  There should be something so I can do `TLSify(someEndpoint, contextFactory)` and get back a new endpoint that does TLS without necessarily doing TCP.

(One example use-case is an endpoint that forwards a port via SSH, or opens a unix socket that some routing software eventually connects to a TCP socket.)

<details><summary>Searchable metadata</summary>

```
trac-id__5642 5642
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_tls_endpoint_wrapper_5642_5 branches/tls-endpoint-wrapper-5642-5
branch_author__hawkowl__habnabit__glyph hawkowl, habnabit, glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1334260183000000 1334260183000000
changetime__1455184020426753 1455184020426753
version__None None
owner__glyph glyph
cc__habnabit cc__ashfall cc__oberstet cc__posita
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to @ashfall|
|-|-|

Since you're working on endpoints.  (It would be neat, but no worries, no particular high priority.)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_2"></a>|@tomprince commented|
|-|-|

[#6308](https://github.com/twisted/twisted/issues/6308) is a duplicate of this:

> The current way of doing SSL with endpoints involves using listenSSL/connectSSL. This means you need one for each of TCPv4 and TCPv6, and we'll need one for the hostname endpoint too. And, you can't use it with e.g. a SOCKS endpoint.

> Better would be an endpoint (or maybe I mean pair of endpoints, one client one server) that wrap an existing endpoint. Ideally this wouldn't even call transport.startTLS, but rather use the TLS protocol implementation directly.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj commented|
|-|-|

habnabit wrote this:
 * https://github.com/habnabit/txsocksx/blob/parsley/txsocksx/ssl.py

```
20:38 < rwall> https://twistedmatrix.com/pipermail/twisted-web/2013-July/005059.html
20:38 < _habnabit> aha
20:39 < _habnabit> i've written that, actually
20:39 < _habnabit> https://github.com/habnabit/txsocksx/blob/parsley/txsocksx/ssl.py
20:39 < rwall> hehe
20:39 < _habnabit> composable endpoints!
20:40 < _habnabit> so maybe there should be a ticket to add this to twisted, blocking on the ticket that adds syntax for it
20:40 < _habnabit> or three tickets? add this, add composable endpoint syntax, add a parser plugin for this
20:40 < rwall> Right I get it now.
20:40 < rwall> yes, that sounds reasonable
20:41 < _habnabit> i think the reason i haven't submitted this to twisted is i have absolutely no idea how to test it
20:42 < _habnabit> somehow having such a small amount of code makes it hard
20:46 < rwall> I don't know it that well, but I guess you pass in a MemoryReactor to the inner endpoint...for a start...and check that you get back a 
               fake port with a tls protocol attached.
20:51 < dreid> How is the tls.TLSMemoryBIOFactory tested?
20:51 < _habnabit> let's see
20:54 < _habnabit> dang, there's a bunch of stuff here
20:55 < _habnabit> it looks like some of the tests involve hooking up two of them to each other--one as a server and the other as a client

```

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj commented|
|-|-|

A discussion about an endpoint string syntax for TLS endpoint wrappers

```
19:51 < rwall> idnar: I got systemd socket activation working with TLS: https://gist.github.com/wallrj/5911925
19:56 < tomprince> It'd be nice if there was an endpoint string parser that would do that.
19:58 < _habnabit> you mean have the ability to wrap endpoints in other endpoints?
19:59 < tomprince> Yes.
20:00 < _habnabit> i was kind of working on this at the pycon sprint last year; i can't remember why it stalled
20:00 < rwall> tomprince: Something like "systemd+tls:domain=INET:index=0" as the string?
20:00 < tomprince> tls:endpoint=systemd\:index\=0:cert=...
20:01 < _habnabit> that gets nasty to write pretty quickly
20:01 < _habnabit> the syntax i proposed was something like 'systemd:domain=INET:index=0;tls'
20:01 < tomprince> tls:endpoint=(systemd:index=0):cert=....
20:02 < _habnabit> hah
20:02 < _habnabit> that would mean making the endpoint string parser a bit more complicated
20:02 < tomprince> More importantly, incompatible I think.
20:03 < tomprince> Anyway, we have parsley, for writing the parser.
20:03 < _habnabit> sure
20:03 < _habnabit> ; would be incompatible too though, hmm
20:03 < _habnabit> : and = are the only reserved characters, so, :: ?
20:04 < rwall> What about IPv6 addresses?
20:04 < _habnabit> they have to be escaped now anyway
20:04 < rwall> They have to be escaped anyway I believe
20:04 < rwall> right
20:05 < _habnabit> maybe that's a good reason to allow () or something as quoting
20:06 < tomprince> I think you definitely want something that allows nesting

```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

This ticket should not be held up by string syntax. That's a separate feature (please open a ticket for it).

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_6"></a>|@habnabit commented|
|-|-|

I have a slightly better and tested version of this in txsocksx now: https://github.com/habnabit/txsocksx/blob/dadcf0567f18f47ffb6d2c0e551b22c0977c0336/txsocksx/ssl.py#L26-62

My previous implementation used `TLSMemoryBIOFactory`, which was not compatible with e.g. `Agent` because of `TLSMemoryBIOFactory`'s use of `WrappingProtocol`. Specifically, `WrappingProtocol` proxies unknown attribute accesses to the *transport* instead of the *wrapped factory*, and `Agent` was trying to access the `request` method on the `Protocol` it was handed back from the endpoint. Normally this would be a `HTTP11ClientProtocol`, but when using `TLSMemoryBIOFactory`, it was instead a `WrappingProtocol`. `wrappingProtocol.request` then resulted in accessing the `request` attribute of a `tcp.Client` instance, which failed.

Instead, the new implementation calls `startTLS` on the protocol's transport as soon as the endpoint's deferred fires with a protocol. This new approach is easier to test, certainly. It won't, however, AFAIK allow for TLS over TLS. I suppose ultimately the question is then "should the resulting endpoint wrapper support TLS over TLS?"

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

`startTLS` is one of those "it would be nice to get rid of it eventually (but it's pretty old and core, who knows if we'll actually ever do that)" APIs.  Still, if `twisted.protocols.tls` is deficient in some way compared to `startTLS`, let's discuss those deficiencies as bugs to be fixed.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Yeah, let's at least try not to add _new_ functionality that depends on `startTLS` unless it needs to.  In the absence of generalized protocol switching, I think some things do still need to; but this does not seem to be one of them, since it starts the protocol as TLS.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

I am a bit skeptical this can't be done using `twisted.protocols.tls`, given that that's what `startTLS` uses... but maybe there's some functionality elsewhere that needs to be used as glue.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_11"></a>|@habnabit commented|
|-|-|

Replying to [itamar](#note_10):
> I am a bit skeptical this can't be done using `twisted.protocols.tls`, given that that's what `startTLS` uses... but maybe there's some functionality elsewhere that needs to be used as glue.

It depends on what you're referring when you say "this". Right now it's certainly _possible_ to use `twisted.protocols.tls` as a wrapper—the difficult part is that then the semantics differ. Specifically, wrapping vs. not-wrapping changes what you get when your endpoint fires with a `Protocol`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

Replying to [habnabit](#note_11):
> Replying to [itamar](#note_10):

> It depends on what you're referring when you say "this". Right now it's certainly _possible_ to use `twisted.protocols.tls` as a wrapper—the difficult part is that then the semantics differ. Specifically, wrapping vs. not-wrapping changes what you get when your endpoint fires with a `Protocol`.

The semantics from the observer's perspective really ought not to differ.  Of course what you get for your `Protocol`'s ''`transport`'' attribute might differ between these two implementation approaches, but you can very easily change what you get back from an endpoint by adding a callback to the (wrapped) endpoint that returns some other value.

Why would it not be possible to add such a callback in this case?

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_13"></a>|@habnabit commented|
|-|-|

Replying to [glyph](#note_12):
> The semantics from the observer's perspective really ought not to differ.  Of course what you get for your `Protocol`'s ''`transport`'' attribute might differ between these two implementation approaches, but you can very easily change what you get back from an endpoint by adding a callback to the (wrapped) endpoint that returns some other value.
> 
> Why would it not be possible to add such a callback in this case?

Ah! That would do it. I simply hadn't thought of that as an option. I'll try it now, but I suspect that's all that's necessary.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_14"></a>|@habnabit commented|
|-|-|

Yes! That is all that's necessary for using `TLSMemoryBIOFactory` as a wrapper. Here's a minimal implementation:

```
class TLSWrapClientEndpoint(object):
    implements(interfaces.IStreamClientEndpoint)

    def __init__(self, contextFactory, wrappedEndpoint):
        self.contextFactory = contextFactory
        self.wrappedEndpoint = wrappedEndpoint

    def connect(self, fac):
        fac = tls.TLSMemoryBIOFactory(self.contextFactory, True, fac)
        return self.wrappedEndpoint.connect(fac).addCallback(self._unwrapProtocol)

    def _unwrapProtocol(self, proto):
        return proto.wrappedProtocol
```

The problem remains that I'm unsure how to test this—is there a fake `TLSMemoryBIOFactory`, or should the test maybe do a real TLS handshake?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

You can test most of this without getting anything TLS-related involved.

"TLS endpoint wrapper" is just a particular instance of "endpoint wrapper".

You can test most of the logic with a much simpler kind of wrapping - say, a wrapper that sends every byte incremented by one.

Then "TLS endpoint wrapper" is just a particular assembly of existing, well-tested components that only needs sufficient tests to verify the assembly was done correctly.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_15):
> You can test most of the logic with a much simpler kind of wrapping - say, a wrapper that sends every byte incremented by one.

This is clearly not a realistic demonstration of a cryptographic protocol.

Now, if you were to ''`xor` every byte'' it with a bit-pattern like `0b01010101` (ASCII: "U") that would totally be a realistic simulation.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_17"></a>|@itamarst commented|
|-|-|

Here are the tests I would write:

1. connect() returns instance of protocol created with given factory.
2. This protocol's transport is a TLSMemwWhatever.
3. This TLSMemwWhatever is using the correct context factory.
4. This TLSMemwWhatever's transport is the transport hooked up by the underlying wrapped endpoint.
5. This TLSMemwWhatever is in client mode.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit set owner to @habnabit<br>@habnabit set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

There's a prototype of something like this in https://github.com/glyph/txsni

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_21"></a>|@habnabit commented|
|-|-|

So, I'm trying to figure out how exactly the clientFromString for this should behave. Originally I wrote a `tls:` string parser that would take an endpoint description and wrap that endpoint.

However, this whole thing, of course, needs hostname validation for clients. It's a bit redundant to specify, say, `tls:tcp\:example.com\:443:hostname=example.com`. I'm thinking now that the `tls:` parser should use `HostnameEndpoint` underneath to establish connections. So, `tls:example.com:443`, and it passes the hostname both to `HostnameEndpoint` and `CertificateOptions`. There can be a separate `tlswrap:` endpoint type that does wrap another endpoint.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_23"></a>|@habnabit commented|
|-|-|

(In [42188]) Branching to 'tls-endpoint-wrapper-5642-2'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit removed owner<br>@habnabit set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_25"></a>|@habnabit commented|
|-|-|

This currently only implements client wrappers; I'm not sure if the branch should implement server wrappers as well at the same time.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_26"></a>|@tomprince commented|
|-|-|

Small branches. Small branches. Small branches.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph commented|
|-|-|

Replying to [tom.prince](#note_26):
> Small branches. Small branches. Small branches.

Yes.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_28"></a>|@glyph set owner to @habnabit|
|-|-|

Thanks very much habnabit.

  1. `TLSWrapperClientEndpoint` should probably just be called `TLSClientEndpoint`.
  2. the `fac` parameter to  should be `protocolFactory`, so sayeth the interface.  (So say we all.)  But also don't have names like "fac".
  3. Pretty sure you don't have to do stuff like this:
```
+    @ivar prefix: See
+        L{IStreamClientEndpointStringParserWithReactor.prefix}.
```
  properties should be the same as methods in this regard. If pydoctor fills in the information correctly like it does with methods, please just remove it.  (Although if you got hassled by twistedchecker or pydoctor about this, please file a bug on the appropriate tool.)
  4. `_parseClient` is very ugly.  This is a bug, or at the very least a missing feature, in `zope.interface`. Obviously it should be left in place to satisfy the `verifyObject` tests or whatever, but please file a bug against ZI and link to it from the docstring.
  5. `TLSWrapperClientEndpoint` is actually a totally general wrapper endpoint - to the point where its test cases actually verify its generality via the upper-casing wrapper factory - made specific only by the fact that it curries only a single argument into its `_wrapper` callable.  If it's any significant amount of work at all, don't bother with this, but it seems like you could trivially convert it into a `WrapperEndpoint` and then just have a wrapTLSEndpoint function that used it with TLSMemoryBIOFactory and did the currying with a lambda rather than something that required documentation as an argument and an ivar and so on.  If you decide not to do this then please file a ticket for generalizing it to a wrapper endpoint to facilitate other kinds of composition.
  6. `test_endpoints.FakeEndpoint`
    1. There are a couple of other bits of code within Twisted which already do parts of this; `twisted.conch.test.test_endpoints.SingleUseMemoryEndpoint`, some stuff in `twisted.internet.test.fakeendpoint`.  As we create fakes like this we should always be careful to clean things up, to centralize them, to test them and to make them public.  Consider putting this somewhere like `proto_helpers`, but at least eliminate the duplication with conch.
    2. Consider splitting the fake into two objects so that the system under test doesn't have access to `endpoint.deferred` without typing at least a few underscores.  It would obviously be bad to end up testing that.  (This is a general design thing we should get better at with test doubles, too.)
  7. Don't create new old-style classes, even in tests :-).  `UppercaseWrapperFactory` and `NetstringTracker` and `NetstringFactory` should all inherit from object.
  8. `NetstringFactory` shouldn't be a `ClientFactory`, endpoints don't even use `ClientFactory`s.
  8. Except don't bother even having `NetstringFactory`.  Kill several lines of docstrings and just use `Factory.forProtocol`.
  9. Assert something useful about the error message in `test_noSSLSupport`.  (Like, that it contains the word "SSL" and ideally something about installing `pyOpenSSL`).  We need to get better about considering the usability of such messages.
  10. `test_utf8Encoding`'s docstring is a little confusing because it uses the word "or".  It passes the correct type to each underlying API, it should just say what it passes where.
  11. `test_defaultSSLOptions` reveals some unfortunate things about the implementation of this API.  Since the `tls:` prefix is _new_ code, we don't have any compatibility expectations for it, so it should be as secure as possible by default.  That means that, being client TLS, it should always verify by default, using the default `platformTrust`.  Turning off verification, while it should still be possible, should be ugly and highly visible on the command line.  `absolutelyNoSecurity:True` is better than `verify:False`.  This means you can't use the `caCerts` and `verify` parameters (which should probably be deprecated anyway) - use the `trustRoot` argument to `CertificateOptions`.
  12. As a corrolary to the previous point, `tlswrap:` should take a _mandatory_ hostname as its first parameter, which can optionally be like `*` or something if you really really don't want to pass a hostname.  (This is extra nice because shell globbing will helpfully screw it up and you'll have to quote the whole thing in order to get it to work.)
  13. Given that `tlswrap:` is a bit more esoteric and is also a bunch of additional code, it would make sense to split this up.  Especially given the length of this review ;-).
  14. In the `endpoints.rst`:
    1. under "TLS", `contextEndpoint` is unused and `contextFactory` is undefined.
    2. in the "TLS wrapper" section, you use `tls:` instead of `tlswrap:` for the example.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_29"></a>|@tomprince commented|
|-|-|

Regarding point 5 of [comment:28 glyph's review]: Given that we eventually want to factor things into a generic wrapper; if it isn't done in this branch, we should avoid exposing the non-generic wrapper endpoint as a public class. That is, just provide `wrapTLSEndpoint` as the public function, and make the class itself private. That way, we won't need to continue providing the non-generic class.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_30"></a>|@hawkowl commented|
|-|-|

(In [45541]) Branching to what-is-ssl-5642.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_32"></a>|@hawkowl commented|
|-|-|

(In [45554]) Branching to tls-endpoint-wrapper-5642-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_33"></a>|@hawkowl commented|
|-|-|

Merged forward, sorted out some merge conflicts.

Any idea what the failures on Py27 are about habnabit? (I'll fix the Py3.4 ones later)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_35"></a>|@glyph commented|
|-|-|

(In [46081]) Branching to tls-endpoint-wrapper-5642-4.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_36"></a>|@glyph commented|
|-|-|

(In [46649]) Branching to tls-endpoint-wrapper-5642-5.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_37"></a>|@glyph removed owner|
|-|-|

Responding to my own review:

>   4. `_parseClient` is very ugly.  This is a bug, or at the very least a missing feature, in `zope.interface`. Obviously it should be left in place to satisfy the `verifyObject` tests or whatever, but please file a bug against ZI and link to it from the docstring.

I'm not even sure what I meant by this at this point.

>  5. `TLSWrapperClientEndpoint` is actually a totally general wrapper endpoint...

habnabit totally addressed this by transitioning to wrapClientTLS.

>  6. `test_endpoints.FakeEndpoint`
>    1. There are a couple of other bits of code within Twisted which already do parts of this; `twisted.conch.test.test_endpoints.SingleUseMemoryEndpoint`, some stuff in `twisted.internet.test.fakeendpoint`.  As we create fakes like this we should always be careful to clean things up, to centralize them, to test them and to make them public.  Consider putting this somewhere like `proto_helpers`, but at least eliminate the duplication with conch.
>    2. Consider splitting the fake into two objects so that the system under test doesn't have access to `endpoint.deferred` without typing at least a few underscores.  It would obviously be bad to end up testing that.  (This is a general design thing we should get better at with test doubles, too.)

I've now deleted FakeEndpoint and created a separated inspector/fake in iosim to assist tests that need an endpoint they can use.  This just uses TCP4ClientEndpoint under the covers, using MemoryReactorClock for verisimilitude.

>  7. Don't create new old-style classes, even in tests :-).  `UppercaseWrapperFactory` and `NetstringTracker` and `NetstringFactory` should all inherit from object.

Done, these now subclass from `object`.

>  8. `NetstringFactory` shouldn't be a `ClientFactory`, endpoints don't even use `ClientFactory`s.
>  9. Except don't bother even having `NetstringFactory`.  Kill several lines of docstrings and just use `Factory.forProtocol`.

I got rid of it.

>  9. Assert something useful about the error message in `test_noSSLSupport`.  (Like, that it contains the word "SSL" and ideally something about installing `pyOpenSSL`).  We need to get better about considering the usability of such messages.

Good thing I had this review comment to look at!  I killed the test when killing the `TLSWrapperClientEndpoint` tests in favor of `wrapClientTLS`.  I've put it back with explicit error checking, along with a utility for checking cases like this that _doesn't_ involve patching global state or coupling to trial's test case like `patch`.

>  10. `test_utf8Encoding`'s docstring is a little confusing because it uses the word "or".  It passes the correct type to each underlying API, it should just say what it passes where.

It took me a few minutes of tracing through the code to remember what this _should_ say, but it's fixed now.

>  11. `test_defaultSSLOptions` reveals some unfortunate things about the implementation of this API.  Since the `tls:` prefix is _new_ code, we don't have any compatibility expectations for it, so it should be as secure as possible by default.  That means that, being client TLS, it should always verify by default, using the default `platformTrust`.  Turning off verification, while it should still be possible, should be ugly and highly visible on the command line.  `absolutelyNoSecurity:True` is better than `verify:False`.  This means you can't use the `caCerts` and `verify` parameters (which should probably be deprecated anyway) - use the `trustRoot` argument to `CertificateOptions`.

I don't think this test exists any more.

>  12. As a corrolary to the previous point, `tlswrap:` should take a _mandatory_ hostname as its first parameter, which can optionally be like `*` or something if you really really don't want to pass a hostname.  (This is extra nice because shell globbing will helpfully screw it up and you'll have to quote the whole thing in order to get it to work.)

I had a much better idea here.  You have to specify a hostname either way ``tls:``.  Rather than using a different endpoint prefix, let's just have an _optional **endpoint** argument_ that transforms from TLS to TLS-wrapper if it's specified.  This is more likely how users are going to think of this anyway.

>  13. Given that `tlswrap:` is a bit more esoteric and is also a bunch of additional code, it would make sense to split this up.  Especially given the length of this review ;-).

Oh well; so much for that :-).

>  14. In the `endpoints.rst`:
>    1. under "TLS", `contextEndpoint` is unused and `contextFactory` is undefined.

Fixed.

>    2. in the "TLS wrapper" section, you use `tls:` instead of `tlswrap:` for the example.

Fixed by the above realization for point 12.

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_38"></a>|@ashfall commented|
|-|-|

Thank you for taking this up. It's great to see progress!

At an initial glance, I have some thoughts on `_matchingString`.

1. `_matchingString` doesn’t seem like a very insightful name (I had to go look at the method definition to understand what it does). Since this is really matching the string “type” (or encoding), I would like the name to indicate that. Unsure if `_matchingStringType`?
2. The actual purpose of  the method is not very easy to find in these 9 lines. 
```
L{_matchingString} will take a constant string (either L{bytes} or L{unicode}) and convert it to the same type as the input string. 
```
was pretty much what I wanted to know first, when looking at the doctsring. I would love it if it could scream that at me.
Do you think you could refactor this to state what `_matchingString` does first, before following it up with the need to implement it (i.e. the `os.path.join` example). Bonus points if you could actually illustrate its use, but since it’s meant to be private, I will not insist on it.
3. Also, consider perhaps:
```
    if type(constantString) == type(inputString):
        return constantString

    if isinstance(constantString, bytes):
        otherType = constantString.decode("ascii")
    else:
        otherType = constantString.encode("ascii”)
    return otherType
```

Oh and, thanks for updating the howto. Really.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_39"></a>|@hawkowl set owner to @glyph|
|-|-|

T GLYPH I'M REVIEWING THIS BUT BECAUSE THIS IS A HARD BRANCH TO REVIEW, I AM FALLING BACK ON `________` HUMOUR AS A COPING MECHANISM FOR THE +900/-68 DIFF.

- IN ENDPOINTS.RST, THERE'S SPACES AFTER THE ENDING `````}}S. THE SPACES ELSEWHERE ARE A SIDE EFFECT OF LORE2RST, SO THEY ARE NO LONGER REQUIRED.

- IN ENDPOINTS.RST, L152, OPTIONAL ARGUMENTS NEED TO BE ON A NEW LINE.

- IN ENDPOINTS.RST, L160, THAT EXAMPLE SEEMS TO BE IN THE WRONG PLACE / DOESN'T MAKE SENSE IN THE UNIX SECTION?

- IN ENDPOINTS.RST, MAYBE WE NEED A RST DIRECTIVE? LIKE, {{{.. deprecated::```, LIKE ```.. note::```? THAT MIGHT LOOK NICER, BUT THIS IS OPTIONAL.

- IN ENDPOINTS.PY, ```wrapClientTLS``` SAYS SINCE 15.5, NOT 16.0.

- IN ENDPOINTS.PY, ```_parseClientSSLOptions``` HAS FOUR SPACES ABOVE IT, NOT THREE.

- IN TEST_ENDPOINTS.PY, ```test_tlsWithDefaults``` MENTIONS SSL IN ITS DOCSTRING.

- IT NEEDS A NEWSFILE.

I AM FAIRLY HAPPY WITH THIS. WHEN ALL ARE GREEN (OS X NEEDS FIXING BUT IT IS UNRELATED, MAYBE CRYPTO WILL FIX MY BUILDS), PLZ MERGE.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_40"></a>|@glyph commented|
|-|-|

> T GLYPH I'M REVIEWING THIS BUT BECAUSE THIS IS A HARD BRANCH TO REVIEW, I AM FALLING BACK ON `________` HUMOUR AS A COPING MECHANISM FOR THE +900/-68 DIFF.

Sorry.  If it's any consolation, a lot of that diff was test utilities which I made the explicit decision to bite the bullet on and make slightly biggger in this PR so that we make progress towards being able to test changes like this _without_ writing giant piles of half-broken new test fakes every time.

> - IN ENDPOINTS.RST, THERE'S SPACES AFTER THE ENDING `````}}S. THE SPACES ELSEWHERE ARE A SIDE EFFECT OF LORE2RST, SO THEY ARE NO LONGER REQUIRED.

Thanks for spotting that.  Fixed.

> - IN ENDPOINTS.RST, L152, OPTIONAL ARGUMENTS NEED TO BE ON A NEW LINE.

Arguably should have been in a new paragraph too.  Fixed.

> - IN ENDPOINTS.RST, L160, THAT EXAMPLE SEEMS TO BE IN THE WRONG PLACE / DOESN'T MAKE SENSE IN THE UNIX SECTION?

I think you meant "L194"?  You're right though, that was a mispaste.

> - IN ENDPOINTS.RST, MAYBE WE NEED A RST DIRECTIVE? LIKE, {{{.. deprecated::```, LIKE ```.. note::```? THAT MIGHT LOOK NICER, BUT THIS IS OPTIONAL.

There doesn't seem to be a `.. deprecated::` directive, so I've switched to `.. note::`.

> - IN ENDPOINTS.PY, ```wrapClientTLS``` SAYS SINCE 15.5, NOT 16.0.

T HAWKOWL SO WHEN ARE WE GETTING `$NEXTVERSION` IN OUR RELEASE AUTOMATION HUH

> - IN ENDPOINTS.PY, ```_parseClientSSLOptions``` HAS FOUR SPACES ABOVE IT, NOT THREE.

Right.

> - IN TEST_ENDPOINTS.PY, ```test_tlsWithDefaults``` MENTIONS SSL IN ITS DOCSTRING.

Fixed.

> - IT NEEDS A NEWSFILE.

Done.

> I AM FAIRLY HAPPY WITH THIS. WHEN ALL ARE GREEN (OS X NEEDS FIXING BUT IT IS UNRELATED, MAYBE CRYPTO WILL FIX MY BUILDS), PLZ MERGE.

Thanks for the review.  I noticed a couple of other minor issues with docstrings and fixed them as well.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_41"></a>|@glyph set status to `closed`|
|-|-|

(In [46774]) Merge tls-endpoint-wrapper-5642-5: "tls:" endpoint

Author: habnabit, glyph

Reviewer: glyph, hawkowl

Fixes: [#5642](https://github.com/twisted/twisted/issues/5642)

Add a new, better "tls:" endpoint, and numerous test utilities to facilitate covering it.
